platforms/core-configuration/kotlin-dsl/src/main/kotlin/org/gradle/kotlin/dsl/provider/KotlinScriptClassPathProvider.kt

```diff
@@ -64,34 +67,58 @@ fun kotlinScriptClassPathProviderOf(project: Project) =
     project.serviceOf<KotlinScriptClassPathProvider>()
 
 
-internal
-typealias JarsProvider = () -> Collection<File>
-
-
 @ServiceScope(Scope.Build::class)
 class KotlinScriptClassPathProvider(
-    private val moduleRegistry: ModuleRegistry,
-    private val classPathRegistry: ClassPathRegistry,
-    private val coreAndPluginsScope: ClassLoaderScope,
-    private val gradleApiJarsProvider: JarsProvider,
+    gradleProperties: GradleProperties,
+    moduleRegistry: ModuleRegistry,
+    dependencyFactory: DependencyFactory,
+    classPathRegistry: ClassPathRegistry,
+    coreAndPluginsScope: ClassLoaderScope,
 ) : GradleApiClasspathProvider {
 
-    /**
-     * Generated Gradle API jar plus supporting libraries such as groovy-all.jar and generated API extensions.
-     */
     private
-    val gradleKotlinDslClasspath: ClassPath by lazy {
-        gradleApi + gradleApiExtensions + gradleKotlinDslJars
+    val gradleAbiClasspath: ClassPath by lazy {
+        moduleRegistry.findModule("gradle-public-api-legacy")
+            ?.allRequiredModulesClasspath
+            ?.asFiles
+            ?.let { files -> DefaultClassPath.of(files) }
+            ?.let { cp ->
+                // Declarative Gradle is not part of the public api yet
+                // Only add it to the compilation classpath if enabled
+                when (gradleProperties.isDclEnabled) {
+                    true -> cp + moduleRegistry.getModule("gradle-project-features").classpath
+                    false -> cp
+                }
+            }
+        // If the distro doesn't contain the ABI jar, fallback to the generated API jar
```

Comment:
Eventually it might make sense for there not to be one gradle API jar, but for each distribution to have their own API jar that it generates and embeds on its own

Reply:
I see this is mentioned in the PR description